File: 143. Reorder List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode node) {
        ListNode prev = null;
        ListNode curr = node;
        while (curr != null) {
            if (curr.next == null) {
                curr.next = prev;
                return curr;
            }
            ListNode temp = curr.next;
            curr.next = prev;
            prev = curr;
            curr = temp;
        }
        return null;
    }

    public void reorderList(ListNode head) {
        int nodes = 0;
        ListNode tort = head;
        ListNode hare = head;
        ListNode curr = head;
        while(curr != null){
            nodes++;
            curr = curr.next;
        }
        while(hare != null && hare.next != null && hare.next.next != null){
            tort = tort.next;
            hare = hare.next.next;
        }
        // System.out.println(tort.val);
        if(nodes % 2 == 0) tort = tort.next;
        tort.next = reverseList(tort.next);
        curr = head;

        ListNode cutOff = tort;
        tort = tort.next;
        System.out.println(nodes);
        while(tort != null){
            ListNode temp = curr.next;
            ListNode tortTemp = tort.next;
            curr.next = tort;
            tort.next = temp;
            curr = temp;
            tort = tortTemp;
        }
        cutOff.next = null;
    }
}

File: 143. Reorder List.cpp
/**
 * Definition for singly-linked list.
 */
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* reverseList(ListNode* node) {
        ListNode* prev = nullptr;
        ListNode* curr = node;
        while (curr != nullptr) {
            if (curr->next == nullptr) {
                curr->next = prev;
                return curr;
            }
            ListNode* temp = curr->next;
            curr->next = prev;
            prev = curr;
            curr = temp;
        }
        return nullptr;
    }

    void reorderList(ListNode* head) {
        if (!head || !head->next) return;

        int nodes = 0;
        ListNode* curr = head;
        while (curr != nullptr) {
            nodes++;
            curr = curr->next;
        }

        ListNode* tort = head;
        ListNode* hare = head;
        while (hare != nullptr && hare->next != nullptr && hare->next->next != nullptr) {
            tort = tort->next;
            hare = hare->next->next;
        }

        if (nodes % 2 == 0) tort = tort->next;

        tort->next = reverseList(tort->next);

        curr = head;
        ListNode* cutOff = tort;
        tort = tort->next;

        while (tort != nullptr) {
            ListNode* temp = curr->next;
            ListNode* tortTemp = tort->next;

            curr->next = tort;
            tort->next = temp;

            curr = temp;
            tort = tortTemp;
        }
        cutOff->next = nullptr;
    }
};
